src/conn/conn_api.c

```diff
@@ -1436,11 +1440,19 @@ __conn_write_config(
 		WT_ERR_NOTFOUND_OK(ret);
 	}
 
-err:	WT_TRET(fclose(fp));
+	/* Flush to disk and close our handle. */
+	WT_ERR(__wt_fp_close(session, &fp));
 
-	/* Don't leave a damaged file in place. */
-	if (ret != 0)
-		(void)__wt_remove(session, filename);
+	/*
+	 * Rename the file into place: that's an atomic operation so we can be
+	 * confident we never see a partial file.
+	 */
+	WT_ERR(__wt_rename(session, WT_BASECONFIG_SET, WT_BASECONFIG));
+
+err:	WT_TRET(__wt_fp_close(session, &fp));
+
+	/* Discard any damaged temporary file, not required but cleaner. */
+	WT_TRET(__wt_remove_if_exists(session, WT_BASECONFIG_SET));
```

Comment:
I think fflush implies fsync, from http://pubs.opengroup.org/onlinepubs/009695399/functions/fflush.html:

> The functionality described on this reference page is aligned with the ISO C standard. Any conflict between the requirements described here and the ISO C standard is unintentional. This volume of IEEE Std 1003.1-2001 defers to the ISO C standard.

> If stream points to an output stream or an update stream in which the most recent operation was not input, fflush() shall cause any unwritten data for that stream to be written to the file, and the st_ctime and st_mtime fields of the underlying file shall be marked for update.

Is there something I'm missing?